resources/aws/auto_scaling_group.go

```diff
@@ -161,3 +163,85 @@ func (asg *AutoScalingGroup) findExisting() (*autoscaling.Group, error) {
 
 	return autoScalingGroups.AutoScalingGroups[0], nil
 }
+
+// FindLegacy returns true if there is an ASG created for the same cluster name
+// but not being part of a CloudFormation stack.
+func (asg *AutoScalingGroup) FindLegacy() (bool, error) {
+	a, err := asg.findExisting()
+	if IsNotFound(err) {
+		return false, nil
+	} else if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	var keyNameFound, keyClusterFound bool
+	for _, td := range a.Tags {
+		if *td.Key == tagKeyName && *td.Value == asg.Name {
+			keyNameFound = true
+		} else if *td.Key == tagKeyCluster && *td.Value == asg.ClusterID {
+			keyClusterFound = true
+		} else if strings.HasPrefix(*td.Key, "aws:cloudformation:") {
+			return false, nil
+		}
+	}
+
+	return keyNameFound && keyClusterFound, nil
+}
+
+// DetachInstances detaches instances from this ASG and returns their details
+func (asg *AutoScalingGroup) DetachInstances() ([]*autoscaling.Instance, error) {
+	autoScalingGroup, err := asg.findExisting()
+	if err != nil {
+		return []*autoscaling.Instance{}, microerror.Mask(err)
+	}
+
+	for _, i := range autoScalingGroup.Instances {
+		_, err = asg.Client.DetachInstances(&autoscaling.DetachInstancesInput{
+			AutoScalingGroupName:           aws.String(asg.Name),
+			InstanceIds:                    []*string{i.InstanceId},
+			ShouldDecrementDesiredCapacity: aws.Bool(false),
+		})
+		if err != nil {
+			return []*autoscaling.Instance{}, microerror.Mask(err)
+		}
+	}
+
+	return autoScalingGroup.Instances, nil
+}
+
+// AttachInstances returns details of the instances attached to this ASG
+func (asg *AutoScalingGroup) AttachInstances(instances []*autoscaling.Instance) error {
+	// set termination policy to newest first so that the old instances don't get removed
+	terminationPolicy := "NewestInstance"
```

Comment:
I know you were explaining that in the standup but I don't follow.

I understand when you add new instances now, they're terminated immediately.
And then when you change back to default nothing happens. Correct? Why is that desired?

Reply:
Yes, with the default policy the old instances being migrated can be terminated after being attached to the new ASG. Changing the policy to `NewestInstance` makes that the recently created instances (the ones already attached to the CF-controlled ASG) are terminated after the old ones are attached.

I think other approaches are possible, like protecting the instances being transferred, let me know WDYT.

Reply:
Makes sense. Thanks for explanation!